File: examples/node-express/src/services/calculator.js
/** Calculator service with basic arithmetic operations. */

export class CalculatorError extends Error {
  constructor(message) {
    super(message);
    this.name = "CalculatorError";
  }
}

export class DivisionByZeroError extends CalculatorError {
  constructor() {
    super("Cannot divide by zero");
    this.name = "DivisionByZeroError";
  }
}

export class InvalidOperationError extends CalculatorError {
  constructor(operation) {
    super(`Invalid operation: ${operation}. Supported: add, subtract, multiply, divide`);
    this.name = "InvalidOperationError";
  }
}

/**
 * Perform basic arithmetic operation.
 * @param {number} a - First operand
 * @param {number} b - Second operand
 * @param {string} operation - One of 'add', 'subtract', 'multiply', 'divide'
 * @returns {number} Result of the operation
 * @throws {DivisionByZeroError} If dividing by zero
 * @throws {InvalidOperationError} If operation is not supported
 */
export function calculate(a, b, operation) {
  if (typeof a !== "number" || typeof b !== "number") {
    throw new CalculatorError("Operands must be numbers");
  }

  switch (operation) {
    case "add":
      return a + b;
    case "subtract":
      return a - b;
    case "multiply":
      return a * b;
    case "divide":
      if (b === 0) {
        throw new DivisionByZeroError();
      }
      return a / b;
    default:
      throw new InvalidOperationError(operation);
  }
}

File: examples/python-fastapi/src/services/calculator.py
"""Calculator service with basic arithmetic operations."""


class CalculatorError(Exception):
    """Base exception for calculator errors."""

    def __init__(self, message: str):
        self.message = message
        super().__init__(message)


class DivisionByZeroError(CalculatorError):
    """Raised when attempting to divide by zero."""

    def __init__(self):
        super().__init__("Cannot divide by zero")


class InvalidOperationError(CalculatorError):
    """Raised when an invalid operation is requested."""

    def __init__(self, operation: str):
        super().__init__(f"Invalid operation: {operation}. Supported: add, subtract, multiply, divide")


def calculate(a: float, b: float, operation: str) -> float:
    """
    Perform basic arithmetic operation.

    Args:
        a: First operand
        b: Second operand
        operation: One of 'add', 'subtract', 'multiply', 'divide'

    Returns:
        Result of the operation

    Raises:
        DivisionByZeroError: If dividing by zero
        InvalidOperationError: If operation is not supported
    """
    if operation == "add":
        return a + b
    elif operation == "subtract":
        return a - b
    elif operation == "multiply":
        return a * b
    elif operation == "divide":
        if b == 0:
            raise DivisionByZeroError()
        return a / b
    else:
        raise InvalidOperationError(operation)